File: 백준/Silver/1991. 트리 순회/트리 순회.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

N = int(input())
tree = {} #dictionary 사용

for i in range(N):
    root, left, right = input().split()
    tree[root]=[left,right]

def preorder(root): #전위 순회
    if root != '.':
        print(root, end='')
        preorder(tree[root][0])
        preorder(tree[root][1])

def inorder(root): #중위 순회
    if root != '.':
        inorder(tree[root][0])
        print(root, end='')
        inorder(tree[root][1])
    
def postorder(root): #후위 순회
    if root!= '.' :
        postorder(tree[root][0])
        postorder(tree[root][1])
        print(root, end='')
        

preorder('A')
print()
inorder('A')
print()
postorder('A')

File: 백준/Silver/1991. 트리 순회/트리 순회.java
import java.util.Scanner;

public class Main {
    static int[][] tree;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        tree = new int[26][2];
        for (int i = 0; i < N; i++) {
            char root = sc.next().charAt(0);
            char left = sc.next().charAt(0);
            char right = sc.next().charAt(0);
            
            if (left == '.')
                tree[root - 'A'][0] = -1;
            else tree[root - 'A'][0] = left - 'A';

            if (right == '.')
                tree[root - 'A'][1] = -1;
            else tree[root - 'A'][1] = right - 'A';

        }
        preOrder(0);
        System.out.println();
        inOrder(0);
        System.out.println();
        postOrder(0);
    }

    public static void preOrder(int n) {
        if (n == -1) return;

        System.out.print((char) (n + 'A')); //루트 먼저출력 
        preOrder(tree[n][0]); //왼쪽노드 탐색 
        preOrder(tree[n][1]); //오른쪽 노드 탐색

    }

    public static void inOrder(int n) {
        if (n == -1) return;
        inOrder(tree[n][0]); //왼쪽노드 탐색
        System.out.print((char) (n + 'A')); //루트 출력
        inOrder(tree[n][1]); //오른쪽노드 탐색

    }

    public static void postOrder(int n) {
        if (n == -1) return;
        postOrder(tree[n][0]); //왼쪽노드 탐색 
        postOrder(tree[n][1]); //오른쪽노드 탐색
        System.out.print((char) (n + 'A')); //루트 출력

    }
}
